### PR TITLE
Support Kafka 2.4

### PR DESCRIPTION
**Pull request by zmstone**
Prior to this PR, the latest API versions supported were for kafka 1.1
This is not a backward compatible upgrade because of the changes in
request & response struct field names.

### CODE REVIEW

**Comment by k32** [comment]
Fantastic job! I was thinking about making some non-backward compatible changes in brod as well (rework supervision tree, drop old interface functions, get rid of single-threaded group subscriber, etc.), what do you think about creating a brod 4.0 workbranch, and target these kafka_protocol changes towards it?

File: test/kpro_txn_tests.erl [hunk 1]
```diff
@@ -290,16 +280,21 @@ make_transactional_id() ->
...
 
 rand(N) -> rand() rem N.
 
-rand(Min, Max) ->
-  Min + rand(Max - Min + 1).
-
 bin(I) when is_integer(I) -> integer_to_binary(I);
 bin(Str) -> iolist_to_binary(Str).
 
+%% reutrns a list of supported produce versions and the highest fetch version.
```
Comment (k32) [inline test/kpro_txn_tests.erl]: typo: returns

File: src/kpro_req_lib.erl [hunk 2]
```diff
@@ -153,26 +170,45 @@ fetch(Vsn, Topic, Partition, Offset, Opts) ->
...
+  %% 2. The broker do not have to send a topic/partition response at all
+  %%    if there is not any new messages appended or no metadata change
+  %%    sice the last fetch response.
+  %%    i.e. Send only changed data.
+  %% The default values are to disable fetch session.
   SessionID = maps:get(session_id, Opts, 0),
-  Epoch = maps:get(epoch, Opts, -1),
+  Epoch = maps:get(session_epoch, Opts, -1),
+  %% Leader epoch is returned from topic-partition metadata.
+  %% kafka partition leader make use of this number ot fence consumers with
```
Comment (k32) [inline src/kpro_req_lib.erl]: typo: "ot fence consumers"

**Review by k32 — APPROVED** [review]

**Comment by zmstone** [comment]
> Fantastic job! I was thinking about making some non-backward compatible changes in brod as well (rework supervision tree, drop old interface functions, get rid of single-threaded group subscriber, etc.), what do you think about creating a brod 4.0 workbranch, and target these kafka_protocol changes towards it?

@k32 sorry for the late fix & reply.
Upgrading the `kafka_protocol` dependency to 3.0 (after the merge of this PR) can be a minor version bump
i.e. `brod` does not have to be non backward compatible.
However I'd leave you and the Klarna team to decide if it makes more sense to bump major version there.

**Comment by k32** [comment]
@zmstone Thanks for the fixes. Please ping me when you consider this PR ready to be merged.

**Comment by zmstone** [comment]
@k32 I believe it is now ready to merge.
There might be something missing, but we'll find out and fix when updating `brod` to use this new version.

**Comment by k32** [comment]
Thank you!

**Pull request merged**